FLT/DedekindDomain/FiniteAdeleRing/TensorPi.lean

```diff
@@ -0,0 +1,131 @@
+    rw [DFinsupp.finset_sum_apply, DFinsupp.finset_sum_apply, Finset.sum_apply]
+    congr with k
+    by_cases h : k = j
+    · subst h; simp
+    · simp [of_eq_of_ne _ _ _ h]
+
+end
+
+section
+
+variable (R M : Type*) [CommRing R] [AddCommGroup M] [Module R M] [Module.Free R M]
+  [Module.Finite R M] {ι : Type*} (N : ι → Type*) [∀ i, AddCommGroup (N i)] [∀ i, Module R (N i)]
+
+open TensorProduct
+
+noncomputable def finitefreeModule_tensorPiEquiv :
+    M ⊗[R] (∀ i, N i) ≃ₗ[R] (Module.Free.ChooseBasisIndex R M →₀ R) ⊗[R] (∀ i, N i) :=
+  TensorProduct.congr (Module.Free.repr R M) (LinearEquiv.refl R ((i : ι) → N i))
+
+noncomputable def finsuppLeft_TensorPi_equiv_piTensor :
+    (Module.Free.ChooseBasisIndex R M →₀ R) ⊗[R] (∀ i, N i) ≃ₗ[R]
+      ∀ i, ( Module.Free.ChooseBasisIndex R M →₀ R) ⊗[R] (N i) :=
+  finsuppScalarLeft R (∀i, N i) (Module.Free.ChooseBasisIndex R M) ≪≫ₗ
+    (finsuppLEquivDirectSum R (∀i, N i) (Module.Free.ChooseBasisIndex R M)) ≪≫ₗ
+    directSumPi_equiv_piSum  ≪≫ₗ
+    LinearEquiv.piCongrRight (fun i ↦(finsuppLEquivDirectSum R (N i)
+    (Module.Free.ChooseBasisIndex R M)).symm)
+    ≪≫ₗ  LinearEquiv.piCongrRight (fun i ↦
+     (finsuppScalarLeft R (N i) (Module.Free.ChooseBasisIndex R M)).symm)
+
+noncomputable def tensorPiEquiv_finitefreeModule :
+    (∀ i, (Module.Free.ChooseBasisIndex R M →₀ R) ⊗[R] N i) ≃ₗ[R] ∀ i, (M ⊗[R] N i):=
+  LinearEquiv.piCongrRight (fun i ↦ (LinearEquiv.rTensor (N i) (Module.Free.repr R M).symm))
+
+noncomputable def tensorPi_equiv_piTensor :
+    M ⊗[R] (∀ i, N i) ≃ₗ[R] ∀ i, (M ⊗[R] N i) :=
+  (finitefreeModule_tensorPiEquiv R M N) ≪≫ₗ (finsuppLeft_TensorPi_equiv_piTensor R M N)
+    ≪≫ₗ (tensorPiEquiv_finitefreeModule R M N)
+
+lemma tensorPi_equiv_piTensor_apply (m : M) (n : ∀ i, N i) :
+    tensorPi_equiv_piTensor R M N (m ⊗ₜ n) = fun i ↦ (m ⊗ₜ n i) := by
+  unfold tensorPi_equiv_piTensor
+  simp only [finitefreeModule_tensorPiEquiv, LinearEquiv.trans_apply, congr_tmul,
+    LinearEquiv.refl_apply]
+  let m' := (Module.Free.repr R M) m
+  have hm' : (Module.Free.repr R M).symm m' = m := by simp [m']
+  rw [← hm']
+  simp only [LinearEquiv.apply_symm_apply]
+  induction m' using Finsupp.induction_linear
+  · ext
+    simp
+  · ext i
+    simp_all [add_tmul]
+  · rename_i j r
+    rw [← LinearEquiv.eq_symm_apply]
+    simp only [tensorPiEquiv_finitefreeModule, LinearEquiv.piCongrRight_symm]
+    ext i
+    simp only [LinearEquiv.piCongrRight_apply, LinearEquiv.rTensor_symm_tmul, LinearEquiv.symm_symm,
+      LinearEquiv.apply_symm_apply, m']
+    rw [finsuppLeft_TensorPi_equiv_piTensor]
+    simp only [LinearEquiv.trans_apply, LinearEquiv.piCongrRight_apply]
+    rw [LinearEquiv.symm_apply_eq]
+    ext k
+    rw [finsuppScalarLeft_apply, LinearMap.rTensor_tmul, Finsupp.lapply_apply,
+      TensorProduct.lid_tmul, Finsupp.single_apply, ite_smul, zero_smul, ← Finsupp.single_apply]
+    apply congrFun
+    apply congrArg
+    clear k hm' m' m
```

Comment:
These three lines can be simplified by just using `congr`

```suggestion
    congr
```

Tested it locally so it's safe to commit directly from here.